File: Stack/Stack.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};


struct Node* top = NULL;

// Push (insert at top)
void push(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    newNode->data = data;
    newNode->next = top;
    top = newNode;
}

// Pop (remove from top)
void pop() {
    if (top == NULL) {
        printf("Stack is empty\n");
        return;
    }
    struct Node* temp = top;
    top = top->next;
    free(temp);
}

// Peek (view top element)
void peek() {
    if (top == NULL) {
        printf("Stack is empty\n");
    } else {
        printf("Top element: %d\n", top->data);
    }
}

// Display the stack
void displayStack() {
    struct Node* temp = top;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

// Free all nodes in the stack
void freeStack() {
    while (top != NULL) {
        pop();
    }
}


int main() {
    push(10);
    push(20);
    push(30);
    displayStack();   

    peek();           

    pop();
    displayStack();   

    freeStack();      
    return 0;
}

File: Stack/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:

    def __init__(self):
        self.top = None
        self.size = 0

    def length(self):
        return self.size

    def push(self, val):
        new_node = Node(val)
        new_node.next = self.top
        self.top = new_node
        self.size += 1

    def pop(self):
        if self.is_empty():
            raise ValueError("Stack is Empty")

        pop_val = self.top.value
        self.top = self.top.next
        self.size -= 1
        return pop_val

    def peek(self):
        if self.is_empty():
            raise ValueError("Stack is Empty")
        return self.top.value

    def is_empty(self):
        return self.top is None

    def __str__(self):
        values = []
        current = self.top
        while current:
            values.append(str(current.value))
            current = current.next
        return " -> ".join(values) if values else "Stack is empty"
